server/gokbg3/grails-app/services/org/gokb/UserProfileService.groovy

```diff
@@ -58,8 +60,110 @@ class UserProfileService {
       user_to_delete.delete(flush: true, failOnError: true)
 
       log.debug("Done")
+      result.result = "OK"
+      result.errors = user_to_delete.errors
     } else {
       log.error("Could not find either the user object for deletion (${params.id}) or the placeholder user")
+      result.result = "ERROR"
+      result.errors = user_to_delete.errors
     }
+    return result
+  }
+
+  def update(User user, def data) {
+    def result = ['result': 'OK']
+    def immutables = ['id', 'username', 'enabled', 'accountExpired', 'accountLocked', 'passwordExpired', 'last_alert_check']
+    def errors = []
+    def skippedCG = false
+    def reqBody = data
+
+    if (reqBody && reqBody.id && reqBody.id == user.id) {
+      reqBody.each { field, val ->
+        if (val && user.hasProperty(field)) {
+          // roles have to be treated separately, as they're not a user property
+          if (field != 'curatoryGroups') {
+            if (!immutables.contains(field)) {
+              user."${field}" = val
+            } else {
+              log.debug("Ignoring immutable field ${field}")
+            }
+          } else {
+            if (user.hasRole('ROLE_EDITOR') || user.hasRole('ROLE_ADMIN')) {
+              def curGroups = []
+              val.each { cg ->
+                def cg_obj = null
+
+                if (cg.uuid?.trim()) {
+                  cg_obj = CuratoryGroup.findByUuid(cg.uuid)
+                }
+
+                if (!cg_obj && cg.id) {
+                  cg_obj = cg.id instanceof String ? genericOIDService.resolveOID(cg.id) : CuratoryGroup.get(cg.id)
+                }
+
+                if (cg_obj) {
+                  curGroups.add(cg_obj)
+                } else {
+                  log.debug("CuratoryGroup ${cg} not found!")
+                  errors << ['message': 'Could not find referenced curatory group!', 'baddata': cg]
+                }
+              }
+              log.debug("New CuratoryGroups: ${curGroups}")
+
+              if (errors.size() > 0) {
+                result.message = "There have been errors updating the users curatory groups."
+                result.errors = errors
+                response.setStatus(400)
+              } else {
+                user.curatoryGroups.addAll(curGroups)
+                user.curatoryGroups.retainAll(curGroups)
+              }
+            } else {
+              skippedCG = true
+            }
+          }
+        } else if (field == "roles") {
+          // scan data
+          Set<Role> newRoles = new HashSet<Role>()
+          val.each { value ->
+            Role newRole = Role.findByAuthority(value.authority)
```

Comment:
May want to check by Role ID (value.id) as well.

Reply:
updated according to the suggestion